File: Client/Classes/CombatDSL/Execution/ActionCompiler.cs
using System;
using System.Collections.Generic;

public sealed class SkillCompiler
{
    public ICombatAction Compile(DSLExpr expr)
    {
        switch (expr)
        {
            case Func func:
                return CompileFunc(func);
            case EType effect:
                return CompileEType(effect);
            default:
                throw new NotSupportedException($"Unsupported DSL expression: {expr.GetType().Name}");
        }
        return null;
    }

    private ICombatAction CompileFunc(Func func)
    {
        switch (func)
        {
            case Inflict inflict:
                return CompileInflict(inflict);
                break;

            case Delay delay:
                return CompileDelay(delay);
                break;

            default:
                throw new NotSupportedException($"Unsupported function: {func.GetType().Name}");
        }
        return null;
    }

    private ICombatAction CompileInflict(Inflict inflict)
    {
        return new DSLActions.Inflict(Compile(inflict.EType), inflict.Turns);
    }

    private ICombatAction CompileDelay(Delay delay)
    {
        return new DSLActions.Delay(delay.Turns, delay.Then);
    }

    private static int ResolveTurns(CType ctype)
    {
        return ctype switch
        {
            Turns turns => turns.Amount,
            _ => throw new NotSupportedException($"Unsupported condition type: {ctype.GetType().Name}")
        };
    }

    private ICombatAction CompileEType(EType effect)
    {
        switch (effect)
        {
            case Damage damage:
                return new DSLActions.Damage(damage.Amount);
            case Heal heal:
                return new DSLActions.Heal(heal.Amount);
            case Poison poison:
                return new DSLActions.Poison(poison.Amount);
            case Bleed bleed:
                return new DSLActions.Bleed(bleed.Amount);
            case Comp comp:
                return new DSLActions.Comp(comp.Effects);
            default:
                throw new NotSupportedException($"Unsupported effect type: {effect.GetType().Name}");
        }
        return null;
    }
    

    private void ApplyDamage(int amount)
    {
        Console.WriteLine($"Damage: {amount}");
    }

    private void ApplyHeal(int amount)
    {
        Console.WriteLine($"Heal: {amount}");
    }

    private void ApplyPoison(int amount, int turns)
    {
        Console.WriteLine($"Poison: {amount} for {turns} turns");
    }

    private void ApplyBleed(int amount, int turns)
    {
        Console.WriteLine($"Bleed: {amount} for {turns} turns");
    }
}

File: server/internal/server/combatDSL/actionCompiler.go
package combatDSL

import "fmt"

type SkillCompiler struct {
	dslCompiler *Compiler
}

func NewSkillCompiler(dslCompiler *Compiler) *SkillCompiler {
	if dslCompiler == nil {
		dslCompiler = NewCompiler(nil)
	}

	return &SkillCompiler{
		dslCompiler: dslCompiler,
	}
}

func (c *SkillCompiler) CompileText(text string) (CombatAction, error) {
	expr, err := c.dslCompiler.CompileText(text)
	if err != nil {
		return nil, err
	}

	return c.Compile(expr)
}

func (c *SkillCompiler) Compile(expr DSLExpr) (CombatAction, error) {
	switch e := expr.(type) {
	case FuncExpr:
		return c.compileFunc(e)
	case EType:
		return c.compileEType(e)
	default:
		return nil, fmt.Errorf("unsupported DSL expression: %T", expr)
	}
}

func (c *SkillCompiler) compileFunc(fn FuncExpr) (CombatAction, error) {
	switch f := fn.(type) {
	case *InflictExpr:
		return c.compileInflict(f)
	case *DelayExpr:
		return c.compileDelay(f)
	default:
		return nil, fmt.Errorf("unsupported function: %T", fn)
	}
}

func (c *SkillCompiler) compileInflict(inflict *InflictExpr) (CombatAction, error) {
	effect, err := c.Compile(inflict.Effect)
	if err != nil {
		return nil, err
	}

	return NewInflictAction(effect, inflict.Turns), nil
}

func (c *SkillCompiler) compileDelay(delay *DelayExpr) (CombatAction, error) {
	return NewDelayAction(delay.Turns, delay.Then), nil
}

func (c *SkillCompiler) compileEType(effect EType) (CombatAction, error) {
	switch e := effect.(type) {
	case *DamageExpr:
		return NewDamageAction(e.Amount), nil
	case *HealExpr:
		return NewHealAction(e.Amount), nil
	case *PoisonExpr:
		return NewPoisonAction(e.Amount), nil
	case *BleedExpr:
		return NewBleedAction(e.Amount), nil
	case *CompositeExpr:
		actions := make([]CombatAction, 0, len(e.Effects))

		for _, effect := range e.Effects {
			action, err := c.Compile(effect)
			if err != nil {
				return nil, err
			}

			actions = append(actions, action)
		}

		return NewCompositeAction(actions), nil
	default:
		return nil, fmt.Errorf("unsupported effect type: %T", effect)
	}
}
